File: src/programmers/p1832/Solution_Jeonghyun.java
//https://school.programmers.co.kr/learn/courses/30/lessons/1832?language=java
//500×500이면 순수 격자 경로 수만 C(998, 499) -> DP

// 0: free
// 1: x
// 2: 이전에 온 방향으로 직진만 가능
// m: row, n: col



class Solution {
    static int MOD = 20170805;
    static int[][] R; // 오른쪽
    static int[][] D; // 아래로
        
    public int solution(int m, int n, int[][] cityMap) {
        int answer = 0;
        
        R = new int[m][n];
        D = new int[m][n];
        
        // 시작
        R[0][0] = 1;
        
        for(int row = 0; row < m; row++){
            for(int col = 0; col < n; col++){
                if(row == 0 && col == 0) continue; // 초기값 보존
                if(cityMap[row][col] == 1) continue; // 못 가는 길
                
                // R(오른쪽) 계산
                if(col > 0){
                    if(cityMap[row][col-1] == 2){ // 직진만 가능
                        R[row][col] = R[row][col-1];
                    }else{
                        R[row][col] = (R[row][col-1] + D[row][col-1]) % MOD;
                    }
                }
                
                // D(아래로) 계산
                if(row > 0){
                    if(cityMap[row-1][col] == 2){ // 직진만 가능
                        D[row][col] = D[row-1][col];
                    }else{
                        D[row][col] = (R[row-1][col] + D[row-1][col]) % MOD;
                    }
                }
            }
        }
        
        answer = (R[m-1][n-1] + D[m-1][n-1]) % MOD;
        return answer;
    }
}

// 채점을 시작합니다.
// 정확성  테스트
// 테스트 1 〉	통과 (37.85ms, 114MB)
// 채점 결과
// 정확성: 100.0
// 합계: 100.0 / 100.0

File: src/programmers/p150369/Solution_Jeonghyun.py
# https://school.programmers.co.kr/learn/courses/30/lessons/150369

def solution(cap, n, deliveries, pickups):
    answer = 0
    d = 0 # 배달 장부
    p = 0 # 수거 장부
    
    for i in range(n - 1, -1, -1):
        # 집의 상자를 장부에 더하기
        d += deliveries[i]
        p += pickups[i]
        
        # 둘 중 하나가 양수면 트럭 필요
        while d > 0 or p > 0:
            answer += ((i+1) * 2)
            d -= cap
            p -= cap
    return answer

# 채점을 시작합니다.
# 정확성  테스트
# 테스트 1 〉	통과 (0.00ms, 11.3MB)
# 테스트 2 〉	통과 (0.00ms, 11.4MB)
# 테스트 3 〉	통과 (0.01ms, 11.6MB)
# 테스트 4 〉	통과 (0.01ms, 11.6MB)
# 테스트 5 〉	통과 (0.01ms, 11.5MB)
# 테스트 6 〉	통과 (0.01ms, 11.5MB)
# 테스트 7 〉	통과 (0.10ms, 11.4MB)
# 테스트 8 〉	통과 (0.19ms, 11.4MB)
# 테스트 9 〉	통과 (0.84ms, 11.5MB)
# 테스트 10 〉	통과 (0.74ms, 11.4MB)
# 테스트 11 〉	통과 (0.33ms, 11.5MB)
# 테스트 12 〉	통과 (0.31ms, 11.7MB)
# 테스트 13 〉	통과 (0.29ms, 11.4MB)
# 테스트 14 〉	통과 (0.26ms, 11.4MB)
# 테스트 15 〉	통과 (5.38ms, 13.3MB)
# 테스트 16 〉	통과 (139.22ms, 13MB)
# 테스트 17 〉	통과 (18.10ms, 13.2MB)
# 테스트 18 〉	통과 (9.73ms, 13.2MB)
# 테스트 19 〉	통과 (8.60ms, 13.1MB)
# 테스트 20 〉	통과 (9.52ms, 13.3MB)
# 채점 결과
# 정확성: 100.0
# 합계: 100.0 / 100.0